File: jclasslib/src/org/gjt/jclasslib/browser/detail/attributes/LineNumberTableAttributeDetailPane.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail.attributes;

import org.gjt.jclasslib.browser.*;
import org.gjt.jclasslib.structures.*;
import org.gjt.jclasslib.structures.attributes.*;

/**
    Detail pane showing a <tt>LineNumberTable</tt> attribute.

    @author <a href="mailto:jclasslib@gmx.net">Ingo Kegel</a>
    @version $Revision: 1.2 $ $Date: 2001-05-31 13:15:25 $
*/
public class LineNumberTableAttributeDetailPane extends AbstractAttributeListDetailPane {

    public LineNumberTableAttributeDetailPane(BrowserServices services) {
        super(services);
    }
    
    protected AbstractAttributeTableModel createTableModel(AttributeInfo attribute) {
        return new AttributeTableModel(attribute);
    }
    
    private class AttributeTableModel extends AbstractAttributeTableModel {
        
        private static final int COLUMN_COUNT = BASE_COLUMN_COUNT + 2;
        
        private static final int START_PC_COLUMN_INDEX = BASE_COLUMN_COUNT;
        private static final int LINE_NUMBER_COLUMN_INDEX = BASE_COLUMN_COUNT + 1;
        
        private static final int LINE_NUMBER_COLUMN_WIDTH = 100;
        
        private LineNumberTableEntry[] lineNumberTable;
        
        public AttributeTableModel(AttributeInfo attribute) {
            super(attribute);
            lineNumberTable = ((LineNumberTableAttribute)attribute).getLineNumberTable();
            initRowNumberStrings();
        }
        
        public int getColumnWidth(int column) {
            switch (column) {
                case START_PC_COLUMN_INDEX:
                    return NUMBER_COLUMN_WIDTH;
                case LINE_NUMBER_COLUMN_INDEX:
                    return LINE_NUMBER_COLUMN_WIDTH;
                default:
                   return NUMBER_COLUMN_WIDTH;
            }
        }
        
        public int getRowCount() {
            return lineNumberTable.length;
        }
        
        public int getColumnCount() {
            return COLUMN_COUNT;
        }
        
        protected String doGetColumnName(int column) {
            switch (column) {
                case START_PC_COLUMN_INDEX:
                   return "start_pc";
                case LINE_NUMBER_COLUMN_INDEX:
                   return "line_number";
                default:
                   return "";
            }
        }
        
        protected Class doGetColumnClass(int column) {
            return Number.class;
        }
        
        protected Object doGetValueAt(int row, int column) {

            LineNumberTableEntry lineNumberTableEntry = lineNumberTable[row];
            
            switch (column) {
                case START_PC_COLUMN_INDEX:
                    return String.valueOf(lineNumberTableEntry.getStartPc());
                case LINE_NUMBER_COLUMN_INDEX:
                    return String.valueOf(lineNumberTableEntry.getLineNumber());
                default:
                    return "";
            }
        }
    }
}

File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/attributes/LineNumberTableAttributeDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail.attributes

import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.browser.detail.TableDetailPane
import org.gjt.jclasslib.structures.attributes.LineNumberTableAttribute
import org.gjt.jclasslib.structures.attributes.LineNumberTableEntry
import java.util.*

class LineNumberTableAttributeDetailPane(services: BrowserServices) : TableDetailPane<LineNumberTableAttribute>(services) {

    override fun createTableModel(attribute: LineNumberTableAttribute) = AttributeTableModel(attribute)
    override val attributeClass: Class<LineNumberTableAttribute>
        get() = LineNumberTableAttribute::class.java

    protected inner class AttributeTableModel(attribute: LineNumberTableAttribute) : ColumnTableModel<LineNumberTableAttribute>(attribute) {
        override fun buildColumns(columns: ArrayList<Column>) {
            super.buildColumns(columns)
            columns.apply {
                add(object : NumberColumn("Start PC") {
                    override fun createValue(rowIndex: Int): Number = lineNumberTable[rowIndex].startPc
                })
            }
            columns.apply {
                add(object : NumberColumn("Line Number", 100) {
                    override fun createValue(rowIndex: Int): Number = lineNumberTable[rowIndex].lineNumber
                })
            }
        }

        override fun getRowCount(): Int {
            return lineNumberTable.size
        }

        private val lineNumberTable: Array<LineNumberTableEntry>
            get() = attribute.lineNumberTable
    }
}

